tracer/src/Datadog.Trace.Tools.Analyzers/PublicApiAnalyzer/PublicApiAnalyzer.cs

```diff
@@ -0,0 +1,312 @@
+﻿// <copyright file="PublicApiAnalyzer.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+// Roughly based on https://github.com/dotnet/roslyn-analyzers/blob/4512de66bb6e21c548ab0d5a83242b70969ba576/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs
+#nullable enable
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.PooledObjects;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Datadog.Trace.Tools.Analyzers.PublicApiAnalyzer
+{
+    /// <summary>
+    /// DD002: Incorrect usage of public API
+    ///
+    /// Finds internal usages of APIs specifically marked with the [PublicApi] flag.
+    /// These methods should not be called directly by our library code, only users should invoke them.
+    /// The analyzer enforces that. requirement
+    ///
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class PublicApiAnalyzer : DiagnosticAnalyzer
```

Comment:
Just thinking out load, but do you see a requirement where we would want to enforce other attribute usages? If so, do you think the current implementation is flexible enough to be able to support other types as well? (using the same Call Site technique)

Reply:
> do you see a requirement where we would want to enforce other attribute usages in the future

You mean that we don't want to call methods decorated with a different attribute, e.g. `[Obsolete]`? If so, yes, it would just be a case of adding them to `GetPublicApiAttributes()` 🙂

Reply:
Got it. Cool. The name of the analyzer is `PublicApiAnalyzer` so it seemed too specific but I can now see that it's not.